library/kotlin/src/io/envoyproxy/envoymobile/Client.kt

```diff
@@ -0,0 +1,73 @@
+package io.envoyproxy.envoymobile
+
+import java.nio.ByteBuffer
+
+interface Client {
+  fun startStream(request: Request, streamCallback: StreamCallback): StreamEmitter
+}
+
+interface StreamCallback {
+  fun onHeaders(statusCode: Int, headers: Map<String, List<String>>)
+  fun onData(byteBuffer: ByteBuffer, endStream: Boolean)
+  fun onMetadata(metadata: Map<String, List<String>>, endStream: Boolean)
+  fun onTrailers(trailers: Map<String, List<String>>)
+  fun onError(error: Error)
+
+  fun onCanceled() // TODO: onInterrupt()?
+  fun onCompletion()
+
+}
+
+interface StreamEmitter {
+  fun sendHeaders(headers: Map<String, List<String>>): Boolean
+  fun sendData(byteBuffer: ByteBuffer, endStream: Boolean): Boolean
+  fun sendMetadata(metadata: Map<String, List<String>>, endStream: Boolean): Boolean
```

Comment:
Suggestions from discussions:
- Return the same `StreamEmitter` to allow for chaining semantics
- `throwable` instead of a boolean to better identify where a chain fails
- `isActive` can be used per other comment to check state later

Reply:
Also, an error returned here is not recoverable for a stream (stream can't be used if it throws)